File: PROGRAMMERS/LEVEL_02/Q42888.cpp
//https://programmers.co.kr/learn/courses/30/lessons/42888
//오픈채팅방
#include <iostream>

#include <sstream>
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

vector<string> solution(vector<string> record) {

    const char ENTER = 'E', LEAVE = 'L', CHANGE = 'C';
    unordered_map<string, string> id_map;
    string command, uid, nick;

    // 입장 또는 변경일 때 id_map맵에 <uid, nick> 저장
    for (string line : record)
    {
        stringstream ss(line);
        ss >> command;

        if (command[0] == ENTER || command[0] == CHANGE)
        {
            ss >> uid;
            ss >> nick;
        
            id_map[uid] = nick;
        }
    }

    // 입장 또는 퇴장일 때 answer벡터에 nick + 메시지 저장
    vector<string> answer;
    for (string line : record)
    {
        stringstream ss(line);
        ss >> command;
        ss >> uid;

        if (command[0] == ENTER)
        {
            answer.push_back(id_map[uid] + "님이 들어왔습니다.");
        }
        else if (command[0] == LEAVE)
        {
            answer.push_back(id_map[uid] + "님이 나갔습니다.");
        }
    }

    return answer;
}

int main()
{
    vector<string> record = { "Enter uid1234 Muzi", "Enter uid4567 Prodo","Leave uid1234","Enter uid1234 Prodo","Change uid4567 Ryan" };

    vector<string> rtn = solution(record);
    cout << "answer : " << endl;
    for (string s : rtn)
    {
        cout << s << endl;
    }

    return 0;
}

File: PROGRAMMERS/LEVEL_02/Q42888.java
package PROGRAMMERS.LEVEL_02;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Q42888 {

    public static void main(String[] args) {

        /*
        record
        ["Enter uid1234 Muzi",
        "Enter uid4567 Prodo",
        "Leave uid1234",
        "Enter uid1234 Prodo",
        "Change uid4567 Ryan"]

        result
        ["Prodo님이 들어왔습니다.",
        "Ryan님이 들어왔습니다.",
        "Prodo님이 나갔습니다.",
        "Prodo님이 들어왔습니다."]
         */

        String[] record = {"Enter uid1234 Muzi",
                        "Enter uid4567 Prodo",
                        "Leave uid1234",
                        "Enter uid1234 Prodo",
                        "Change uid4567 Ryan"};
        List<String> res = solution(record);
        for (String s : res) {
            System.out.println(s);
        }
    }

    static final String ENTER = "Enter";
    static final String LEAVE = "Leave";
    static final String CHANGE = "Change";

    public static List<String> solution(String[] record) {

        Map<String, String> actionMap = new HashMap<>();
        actionMap.put(ENTER, "님이 들어왔습니다.");
        actionMap.put(LEAVE, "님이 나갔습니다.");

        Map<String, String> idMap = new HashMap<>();
        List<String> idMsgList = new ArrayList<>();
        List<String> actionMsgList = new ArrayList<>();

        for (String str : record) {
            String[] cmd = str.split("\\s");

            if (cmd[0].equals(ENTER)) {

                if (idMap.containsKey(cmd[1])) {
                    idMap.put(cmd[1], cmd[2]);
                } else {
                    idMap.put(cmd[1], cmd[2]);
                }

                idMsgList.add(cmd[1]);
                actionMsgList.add(actionMap.get(cmd[0]));

            } else if (cmd[0].equals(LEAVE)) {
                idMsgList.add(cmd[1]);
                actionMsgList.add(actionMap.get(cmd[0]));

            } else if (cmd[0].equals(CHANGE)) {
                idMap.put(cmd[1], cmd[2]);
            }
        }

        List<String> answer = new ArrayList<>();
        for (int i = 0; i < idMsgList.size(); ++i) {
            answer.add(idMap.get(idMsgList.get(i)) + actionMsgList.get(i));
        }
        return answer;
    }
}
